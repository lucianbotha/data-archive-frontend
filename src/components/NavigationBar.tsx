import {
  faShoppingCart,
  faSignInAlt,
  faUserPlus
} from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import * as React from "react";
import { NavLink } from "react-router-dom";
import styled from "styled-components";

/**
 * The navigation bar containing links to all pages, as well as a link for
 * logging out and a link to the shopping cart.
 *
 * Properties:
 * -----------
 * user : User
 *     The currently logged in user. If the user is not logged in, a falsy value
 *     must be passed.
 * logout : () => void
 *     A function for logging out.
 */

// TODO: replace user type with correct one
interface INavigationBarProps {
  user?: {
    isAdmin: () => boolean;
    name: string;
    username: string;
  } | null; // currently logged in user
  logout: () => void; // logout function
}

interface INavigationBarState {
  isMenuActive: boolean;
}

const Nav = styled.nav.attrs({
  ariaLabel: "menu",
  className: "navbar is-full-width "
})``;

class NavigationBar extends React.Component<
  INavigationBarProps,
  INavigationBarState
> {
  public state = {
    isMenuActive: false
  };

  public render() {
    // classes for highlighting a link to nan active tab
    const activeTab = "has-text-weight-bold has-text-link";

    const { logout, user } = this.props;
    const { isMenuActive } = this.state;

    return (
      <div>
        <Nav>
          <div className="navbar-brand">
            {/* "Burger" for toggling the menu on a mobile device */}
            <a
              className={`navbar-burger burger ${
                isMenuActive ? "is-active" : ""
              }`}
              role="button"
              aria-label="menu"
              aria-expanded={isMenuActive ? "true" : "false"}
              onClick={this.toggleMenu}
            >
              <span aria-hidden="true" />
              <span aria-hidden="true" />
              <span aria-hidden="true" />
            </a>
          </div>

          {/* Menu */}
          <div
            className={`navbar-menu has-text-weight-light  ${
              isMenuActive ? "is-active" : ""
            }`}
          >
            <div className="navbar-start ">
              {/* Link to search page */}
              <NavLink
                activeClassName={activeTab}
                className="navbar-item item"
                exact={true}
                to="/"
              >
                Search
              </NavLink>

              {/* Link to data requests page */}
              {user && (
                <NavLink
                  activeClassName={activeTab}
                  className="navbar-item item"
                  to="/data-requests"
                >
                  Data Requests
                </NavLink>
              )}

              {/* Link to admin page */}
              {user && user.isAdmin() && (
                <NavLink
                  activeClassName={activeTab}
                  className="navbar-item"
                  to="/admin"
                >
                  Admin
                </NavLink>
              )}
            </div>
            <div className="navbar-end subtitle is-4">
              {/* Dropdown menu for account related content */}
              {user && (
                <div className="navbar-item has-dropdown is-hoverable">
                  <a className="navbar-link">Hello {`${user.name}`}</a>

                  <div className="navbar-dropdown">
                    {/* Link to page for editing account details */}
                    <NavLink
                      activeClassName={activeTab}
                      className="navbar-item"
                      to="/account"
                    >
                      Account
                    </NavLink>

                    {/* Link for logging out */}
                    <NavLink
                      className="navbar-item"
                      data-test="logout"
                      onClick={logout}
                      to="/"
                    >
                      Logout
                    </NavLink>
                  </div>
                </div>
              )}

              {!user && (
                <>
                  {/* Link for logging in */}
                  <div className={"navbar-item"}>
                    <NavLink
                      className=" button is-primary  is-outlined"
                      to="/login"
                    >
                      <span>
                        Login <FontAwesomeIcon icon={faSignInAlt} />
                      </span>
                    </NavLink>
                  </div>

                  {/* Link for registering */}
                  <div className={"navbar-item"}>
                    <NavLink
                      className=" button is-info is-outlined"
                      to="/register"
                    >
                      <span>
                        Register <FontAwesomeIcon icon={faUserPlus} />
                      </span>
                    </NavLink>
                  </div>
                </>
              )}

              {/* Cart icon */}
              <div className={"navbar-item is-hidden-descktop"}>
                <NavLink className="button is-link is-outlined" to="/cart">
                  <span>
                    <FontAwesomeIcon icon={faShoppingCart} /> CART
                  </span>
                </NavLink>
              </div>
            </div>
          </div>
        </Nav>
        <hr className="navbar-divider" />
      </div>
    );
  }

  private toggleMenu = () => {
    this.setState(prevState => ({
      isMenuActive: !prevState.isMenuActive
    }));
  };
}

export default NavigationBar;
