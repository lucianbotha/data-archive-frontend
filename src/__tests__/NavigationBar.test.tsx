import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import * as React from "react";
import { MemoryRouter } from "react-router";
import NavigationBar from "../components/NavigationBar";

const mockUser = (isAdmin: boolean, name: string, username: string) => ({
  isAdmin: () => isAdmin,
  name,
  username
});

describe("NavigationBar", () => {
  it("shows the correct content when the user is not logged in", () => {
    const user = null;
    const logout = jest.fn();
    const wrapper = mount(
      <MemoryRouter>
        <NavigationBar user={user} logout={logout} />
      </MemoryRouter>
    );

    // Login button is shown
    expect(wrapper.exists('Link[to="/login"]')).toBeTruthy();

    // Registration button is shown
    expect(wrapper.exists('Link[to="/register"]')).toBeTruthy();

    // Data requests link is not shown
    expect(wrapper.exists('Link[to="/data-requests"]')).toBeFalsy();

    // Account details link is not shown
    expect(wrapper.exists('Link[to="/account"]')).toBeFalsy();

    // Logout link is not shown
    expect(wrapper.exists('a[data-test="logout"]')).toBeFalsy();

    // Overall content check
    const navigationBar = wrapper.find(".navbar");
    expect(toJSON(navigationBar)).toMatchSnapshot();
  });

  it("shows the correct content when the user is a non-administrator", () => {
    const user = mockUser(false, "name", "username");
    const logout = jest.fn();
    const wrapper = mount(
      <MemoryRouter>
        <NavigationBar user={user} logout={logout} />
      </MemoryRouter>
    );

    // Login button is not shown
    expect(wrapper.exists('Link[to="/login"]')).toBeFalsy();

    // Registration button is not shown
    expect(wrapper.exists('Link[to="/register"]')).toBeFalsy();

    // Data requests link is shown
    expect(wrapper.exists('Link[to="/data-requests"]')).toBeTruthy();

    // Account details link is shown
    expect(wrapper.exists('Link[to="/account"]')).toBeTruthy();

    // Logout link is shown
    expect(wrapper.exists('a[data-test="logout"]')).toBeTruthy();

    // Admin page link is not shown
    expect(wrapper.exists('Link[to="/admin"]')).toBeFalsy();

    // Overall content check
    const navigationBar = wrapper.find(".navbar");
    expect(toJSON(navigationBar)).toMatchSnapshot();
  });

  it("shows the correct content when the user is an administrator", () => {
    const user = mockUser(true, "name", "username");
    const logout = jest.fn();
    const wrapper = mount(
      <MemoryRouter>
        <NavigationBar user={user} logout={logout} />
      </MemoryRouter>
    );

    // Login button is not shown
    expect(wrapper.exists('Link[to="/login"]')).toBeFalsy();

    // Registration button is not shown
    expect(wrapper.exists('Link[to="/register"]')).toBeFalsy();

    // Data requests link is shown
    expect(wrapper.exists('Link[to="/data-requests"]')).toBeTruthy();

    // Account details link is shown
    expect(wrapper.exists('Link[to="/account"]')).toBeTruthy();

    // Logout link is shown
    expect(wrapper.exists('a[data-test="logout"]')).toBeTruthy();

    // Admin page link is shown
    expect(wrapper.exists('Link[to="/admin"]')).toBeTruthy();

    // Overall content check
    const navigationBar = wrapper.find(".navbar");
    expect(toJSON(navigationBar)).toMatchSnapshot();
  });

  it("calls the logout function when the logout link is clicked", () => {
    const user = mockUser(false, "name", "username");
    const logout = jest.fn();
    const wrapper = mount(
      <MemoryRouter>
        <NavigationBar user={user} logout={logout} />
      </MemoryRouter>
    );

    // Click the logout link
    const logoutLink = wrapper.find('a[data-test="logout"]');
    logoutLink.simulate("click");
    expect(logout).toHaveBeenCalled();
  });

  it("toggles the visibility of the menu when the burger button is clicked", async () => {
    const user = mockUser(false, "name", "username");
    const logout = jest.fn();
    const wrapper = mount(
      <MemoryRouter>
        <NavigationBar user={user} logout={logout} />
      </MemoryRouter>
    );

    // The burger button and the menu are not active
    let burger = wrapper.find(".burger");
    let menu = wrapper.find(".navbar-menu");
    expect(burger.hasClass("is-active")).toBeFalsy();
    expect(menu.hasClass("is-active")).toBeFalsy();

    // Click the burger
    burger.simulate("click");

    // The burger button and the menu are active now
    burger = wrapper.find(".burger");
    menu = wrapper.find(".navbar-menu");
    expect(burger.hasClass("is-active")).toBeTruthy();
    expect(menu.hasClass("is-active")).toBeTruthy();

    // Click the burger again
    burger.simulate("click");

    // The burger button and the menu are inactive again
    burger = wrapper.find(".burger");
    menu = wrapper.find(".navbar-menu");
    expect(burger.hasClass("is-active")).toBeFalsy();
    expect(menu.hasClass("is-active")).toBeFalsy();
  });
});
