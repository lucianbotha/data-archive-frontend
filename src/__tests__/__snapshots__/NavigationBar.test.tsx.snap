// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationBar shows the correct content when the user is a non-administrator 1`] = `
<nav
  className="sc-bdVaJa navbar is-light is-full-width bDWFJH"
  role="button"
>
  <div
    className="navbar-brand"
  >
    <a
      aria-expanded="false"
      aria-label="menu"
      className="navbar-burger burger "
      onClick={[Function]}
      role="button"
    >
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
    </a>
  </div>
  <div
    className="navbar-menu title  "
  >
    <div
      className="navbar-start "
    >
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="navbar-item is-1"
        to="/"
      >
        <Route
          path="\\\\/"
        >
          <Link
            aria-current="page"
            className="navbar-item is-1 active"
            replace={false}
            style={Object {}}
            to="/"
          >
            <a
              aria-current="page"
              className="navbar-item is-1 active"
              href="/"
              onClick={[Function]}
              style={Object {}}
            >
              Search
            </a>
          </Link>
        </Route>
      </NavLink>
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="navbar-item is-3"
        to="/data-requests"
      >
        <Route
          path="\\\\/data-requests"
        >
          <Link
            aria-current={null}
            className="navbar-item is-3"
            replace={false}
            to="/data-requests"
          >
            <a
              aria-current={null}
              className="navbar-item is-3"
              href="/data-requests"
              onClick={[Function]}
            >
              Data Requests
            </a>
          </Link>
        </Route>
      </NavLink>
    </div>
    <div
      className="navbar-end subtitle is-4"
    >
      <div
        className="navbar-item has-dropdown is-hoverable"
      >
        <a
          className="navbar-link"
        >
          Hello 
          name
        </a>
        <div
          className="navbar-dropdown"
        >
          <NavLink
            activeClassName="active"
            aria-current="page"
            className="navbar-item"
            to="/account"
          >
            <Route
              path="\\\\/account"
            >
              <Link
                aria-current={null}
                className="navbar-item"
                replace={false}
                to="/account"
              >
                <a
                  aria-current={null}
                  className="navbar-item"
                  href="/account"
                  onClick={[Function]}
                >
                  Account
                </a>
              </Link>
            </Route>
          </NavLink>
          <a
            className="navbar-item"
            data-test="logout"
            onClick={[MockFunction]}
          >
            Logout
          </a>
        </div>
      </div>
      <div
        className="navbar-item is-hidden-descktop"
      >
        <NavLink
          activeClassName="active"
          aria-current="page"
          className=" button is-link"
          to="/cart"
        >
          <Route
            path="\\\\/cart"
          >
            <Link
              aria-current={null}
              className=" button is-link"
              replace={false}
              to="/cart"
            >
              <a
                aria-current={null}
                className=" button is-link"
                href="/cart"
                onClick={[Function]}
              >
                <span>
                  <i
                    className="fas fa-shopping-cart"
                  />
                   CART
                </span>
              </a>
            </Link>
          </Route>
        </NavLink>
      </div>
    </div>
  </div>
</nav>
`;

exports[`NavigationBar shows the correct content when the user is an administrator 1`] = `
<nav
  className="sc-bdVaJa navbar is-light is-full-width bDWFJH"
  role="button"
>
  <div
    className="navbar-brand"
  >
    <a
      aria-expanded="false"
      aria-label="menu"
      className="navbar-burger burger "
      onClick={[Function]}
      role="button"
    >
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
    </a>
  </div>
  <div
    className="navbar-menu title  "
  >
    <div
      className="navbar-start "
    >
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="navbar-item is-1"
        to="/"
      >
        <Route
          path="\\\\/"
        >
          <Link
            aria-current="page"
            className="navbar-item is-1 active"
            replace={false}
            style={Object {}}
            to="/"
          >
            <a
              aria-current="page"
              className="navbar-item is-1 active"
              href="/"
              onClick={[Function]}
              style={Object {}}
            >
              Search
            </a>
          </Link>
        </Route>
      </NavLink>
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="navbar-item is-3"
        to="/data-requests"
      >
        <Route
          path="\\\\/data-requests"
        >
          <Link
            aria-current={null}
            className="navbar-item is-3"
            replace={false}
            to="/data-requests"
          >
            <a
              aria-current={null}
              className="navbar-item is-3"
              href="/data-requests"
              onClick={[Function]}
            >
              Data Requests
            </a>
          </Link>
        </Route>
      </NavLink>
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="navbar-item is-3"
        to="/admin"
      >
        <Route
          path="\\\\/admin"
        >
          <Link
            aria-current={null}
            className="navbar-item is-3"
            replace={false}
            to="/admin"
          >
            <a
              aria-current={null}
              className="navbar-item is-3"
              href="/admin"
              onClick={[Function]}
            >
              Admin
            </a>
          </Link>
        </Route>
      </NavLink>
    </div>
    <div
      className="navbar-end subtitle is-4"
    >
      <div
        className="navbar-item has-dropdown is-hoverable"
      >
        <a
          className="navbar-link"
        >
          Hello 
          name
        </a>
        <div
          className="navbar-dropdown"
        >
          <NavLink
            activeClassName="active"
            aria-current="page"
            className="navbar-item"
            to="/account"
          >
            <Route
              path="\\\\/account"
            >
              <Link
                aria-current={null}
                className="navbar-item"
                replace={false}
                to="/account"
              >
                <a
                  aria-current={null}
                  className="navbar-item"
                  href="/account"
                  onClick={[Function]}
                >
                  Account
                </a>
              </Link>
            </Route>
          </NavLink>
          <a
            className="navbar-item"
            data-test="logout"
            onClick={[MockFunction]}
          >
            Logout
          </a>
        </div>
      </div>
      <div
        className="navbar-item is-hidden-descktop"
      >
        <NavLink
          activeClassName="active"
          aria-current="page"
          className=" button is-link"
          to="/cart"
        >
          <Route
            path="\\\\/cart"
          >
            <Link
              aria-current={null}
              className=" button is-link"
              replace={false}
              to="/cart"
            >
              <a
                aria-current={null}
                className=" button is-link"
                href="/cart"
                onClick={[Function]}
              >
                <span>
                  <i
                    className="fas fa-shopping-cart"
                  />
                   CART
                </span>
              </a>
            </Link>
          </Route>
        </NavLink>
      </div>
    </div>
  </div>
</nav>
`;

exports[`NavigationBar shows the correct content when the user is not logged in 1`] = `
<nav
  className="sc-bdVaJa navbar is-light is-full-width bDWFJH"
  role="button"
>
  <div
    className="navbar-brand"
  >
    <a
      aria-expanded="false"
      aria-label="menu"
      className="navbar-burger burger "
      onClick={[Function]}
      role="button"
    >
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
    </a>
  </div>
  <div
    className="navbar-menu title  "
  >
    <div
      className="navbar-start "
    >
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="navbar-item is-1"
        to="/"
      >
        <Route
          path="\\\\/"
        >
          <Link
            aria-current="page"
            className="navbar-item is-1 active"
            replace={false}
            style={Object {}}
            to="/"
          >
            <a
              aria-current="page"
              className="navbar-item is-1 active"
              href="/"
              onClick={[Function]}
              style={Object {}}
            >
              Search
            </a>
          </Link>
        </Route>
      </NavLink>
    </div>
    <div
      className="navbar-end subtitle is-4"
    >
      <div
        className="navbar-item"
      >
        <NavLink
          activeClassName="active"
          aria-current="page"
          className=" button is-primary  is-outlined"
          to="/login"
        >
          <Route
            path="\\\\/login"
          >
            <Link
              aria-current={null}
              className=" button is-primary  is-outlined"
              replace={false}
              to="/login"
            >
              <a
                aria-current={null}
                className=" button is-primary  is-outlined"
                href="/login"
                onClick={[Function]}
              >
                <span>
                  Login 
                  <i
                    className="fas fa-sign-in-alt"
                  />
                   
                </span>
              </a>
            </Link>
          </Route>
        </NavLink>
      </div>
      <div
        className="navbar-item"
      >
        <NavLink
          activeClassName="active"
          aria-current="page"
          className=" button is-info is-outlined"
          to="/register"
        >
          <Route
            path="\\\\/register"
          >
            <Link
              aria-current={null}
              className=" button is-info is-outlined"
              replace={false}
              to="/register"
            >
              <a
                aria-current={null}
                className=" button is-info is-outlined"
                href="/register"
                onClick={[Function]}
              >
                <span>
                  Register 
                  <i
                    className="fas fa-user-plus"
                  />
                </span>
              </a>
            </Link>
          </Route>
        </NavLink>
      </div>
      <div
        className="navbar-item is-hidden-descktop"
      >
        <NavLink
          activeClassName="active"
          aria-current="page"
          className=" button is-link"
          to="/cart"
        >
          <Route
            path="\\\\/cart"
          >
            <Link
              aria-current={null}
              className=" button is-link"
              replace={false}
              to="/cart"
            >
              <a
                aria-current={null}
                className=" button is-link"
                href="/cart"
                onClick={[Function]}
              >
                <span>
                  <i
                    className="fas fa-shopping-cart"
                  />
                   CART
                </span>
              </a>
            </Link>
          </Route>
        </NavLink>
      </div>
    </div>
  </div>
</nav>
`;
